New option for create: --google-use-internal-ip

emsi opened the pull request:
Introduced a new flag for google driver:
--google-use-internal-ip

When invoked while create it will make docker-machine use internal rather than public NATed IPs.
It's very useful if you manage machines from within the same network. It's faster, simpler and make deploying e.g. swarm much easier as one do not have to configure firewall even when swarm is managed from within the same network. The flag is persistent in the sense that a machine created with it retains the feature.

GordonTheTurtle commented:
Looks like we would not be able to merge this PR because of merge conflicts. Please rebase, fix conflicts, and force push to your branch.

dmp42 commented:
Hey Mr. @dgageot wdyt? :)

dgageot commented:
@dmp42 @emsi Seems cool! I'll comment on a few things that could be improved on the code

dgageot commented:
@emsi can you squash the PR into a single commit?

dgageot commented:
@emsi would you like to use the feedback to prepare the PR for being shipped with 0.5.0?

emsi commented:
Hi

Yes. I've been really busy this week. I'll try to update the PR this weekend. I hope it's not too late  :)

-- 
Sent from my phone 
Mariusz Woloszyn

> On 10 paź 2015, at 15:30, David Gageot <notifications@github.com> wrote:
> 
> @emsi would you like to use the feedback to prepare the PR for being shipped with 0.5.0?
> 
> —
> Reply to this email directly or view it on GitHub.
>

dgageot commented:
@emsi Take your time! I just wanted to be sure you were still involved.

emsi commented:
@dgageot: It's squashed now and I have improved the code based on you suggestions. :)

dgageot commented:
Thanks @emsi! I'll test it tomorrow

dgageot commented:
@emsi Can you update the documentation in [docs/drivers/gce.md](https://github.com/docker/machine/blob/master/docs/drivers/gce.md)?

dgageot commented:
@emsi You might want to rebase on #1973 for your tests because on master, the google driver is broken

emsi commented:
I noticed that yesterday and gave up as it was very late already :) Thanx
for pointing to the right commit.

On Tue, Oct 13, 2015 at 12:04 AM, David Gageot <notifications@github.com>
wrote:

> @emsi <https://github.com/emsi> You might want to rebase on #1973
> <https://github.com/docker/machine/pull/1973> for your tests because on
> master, the google driver is broken
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/machine/pull/1879#issuecomment-147534192>.
>



-- 
Emsi

emsi commented:
@dgageot: I fixed that garbage code and added the documentation. I also retested it on the GCE to make sure everything works as expected. 
You can test it now :)

dgageot commented:
Thanks @emsi I'll test that tomorrow first thing

dgageot commented:
@emsi Thank you very much for your PR. I listed a couple of tiny things to fix and I've tested the feature. When the fixes are there, it's good to be merged!

emsi commented:
It should be OK now. :)

dgageot commented:
LGTM

Thank you @emsi 
@nathanleclaire @dmp42 can you take a second look?

emsi commented:
Please hold on. I found small improvement.

emsi commented:
It's perfect now. I removed a bit of unnecessary code.

dgageot commented:
LGTM

dgageot commented:
@dmp42 @nathanleclaire wdyt?

dmp42 commented:
Code LGTM.
@nathanleclaire ptal

dmp42 commented:
@emsi can you rebase against latest master? with the recent plugins merge, we need to make sure everything compiles against the latest master.
Sorry for the inconvenience.

nathanleclaire commented on drivers/google/google.go:
```diff
@@ -106,6 +107,11 @@ func GetCreateFlags() []cli.Flag {
 			Usage:  "GCE Instance Tags (comma-separated)",
 			EnvVar: "GOOGLE_TAGS",
 		},
+		cli.BoolFlag{
```
Specifically @dmp42 mentions this because the issue is with this flag: it should be of type `mcnflag.BoolFlag` on latest master

emsi replied:
DONE

emsi commented:
I rebased against the latest master but google driver seems to be broken there. 
I got ```Error creating machine: Error in driver during machine creation: unexpected EOF``` even without my patch.
I tried with v0.5.0-rc1 and got the same error. When I tried to debug that issue it looks like the interactive part with opening URL and pasting the code goes wrong in the background process (it clearly does not show on the tty while it's in the debug).

nathanleclaire commented:
Ah, right.  Since the code actually makes a call to open a browser window, which naturally won't work inside of an exec-ed process.

Perhaps we need to fix that code to get the token ahead of time like the other drivers, before the release happens.

dgageot commented:
@nathanleclaire @emsi I created a ticket for this issue #2023

nathanleclaire commented:
Thanks @dgageot

dgageot commented:
Hi @emsi, we didn't forget you. The google driver is being fixed with https://github.com/docker/machine/pull/2087. Once it's merged you will have to rebase and fix the conflicts.

Sorry for the delay.

emsi commented:
No problem. I'll stay tuned :)

-- 
Sent from my phone 
Mariusz Woloszyn

> On 27 paź 2015, at 10:53, David Gageot <notifications@github.com> wrote:
> 
> Hi @emsi, we didn't forget you. The google driver is being fixed with #2087. Once it's merged you will have to rebase and fix the conflicts.
> 
> Sorry for the delay.
> 
> —
> Reply to this email directly or view it on GitHub.
>

dgageot commented:
@emsi You should be able to rebase now that the driver is ok \O/

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "google-use-internal-ip" git@github.com:emsi/machine.git somewhere
$ cd somewhere
$ git rebase -i HEAD~2
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

emsi commented:
@dgageot It should be OK now :)

dgageot commented:
LGTM. Scheduling for 0.5.1

jeanlaurent commented:
LGTM

dgageot commented:
Hi @emsi, one last rebase and it's merged! Sorry for the inconvenience. A lot of PRs touched the gce driver at the same time.

emsi commented:
@dgageot: I take your word for it :)
Should be OK now.

dgageot commented:
I think you deserve a medal for all how many times you've rebased your code! :gift: :gift: :gift:

emsi commented:
Great! :)
I got the black belt in rebasing.

dgageot commented:
:+1:
